watchtower/wtclient/session_negotiator.go

```diff
@@ -0,0 +1,453 @@
+	if !ok {
+		return fmt.Errorf("expected Init, got %T in reply", remoteMsg)
+	}
+
+	// Verify the watchtower's remote Init message against our own.
+	err = n.localInit.CheckRemoteInit(remoteInit, wtwire.FeatureNames)
+	if err != nil {
+		return err
+	}
+
+	policy := n.cfg.Policy
+	createSession := &wtwire.CreateSession{
+		BlobType:     policy.BlobType,
+		MaxUpdates:   policy.MaxUpdates,
+		RewardBase:   policy.RewardBase,
+		RewardRate:   policy.RewardRate,
+		SweepFeeRate: policy.SweepFeeRate,
+	}
+
+	// Send CreateSession message.
+	err = n.cfg.SendMessage(conn, createSession)
+	if err != nil {
+		return fmt.Errorf("unable to send CreateSession: %v", err)
+	}
+
+	// Receive CreateSessionReply message.
+	remoteMsg, err = n.cfg.ReadMessage(conn)
+	if err != nil {
+		return fmt.Errorf("unable to read CreateSessionReply: %v", err)
+	}
+
+	// Check that returned message is wtwire.CreateSessionReply.
+	createSessionReply, ok := remoteMsg.(*wtwire.CreateSessionReply)
+	if !ok {
+		return fmt.Errorf("expected CreateSessionReply, got %T in "+
+			"reply", remoteMsg)
+	}
+
+	switch createSessionReply.Code {
+	case wtwire.CodeOK, wtwire.CreateSessionCodeAlreadyExists:
+
+		// TODO(conner): add last-applied to create session reply to
+		// handle case where we lose state, session already exists, and
+		// we want to possibly resume using the session
+
+		// TODO(conner): validate reward address
+		rewardPkScript := createSessionReply.Data
+
+		sessionID := wtdb.NewSessionIDFromPubKey(
+			privKey.PubKey(),
+		)
+		clientSession := &wtdb.ClientSession{
+			TowerID:        tower.ID,
+			Tower:          tower,
+			SessionPrivKey: privKey, // remove after using HD keys
+			ID:             sessionID,
+			Policy:         n.cfg.Policy,
+			SeqNum:         0,
+			RewardPkScript: rewardPkScript,
+		}
+
+		err = n.cfg.DB.CreateClientSession(clientSession)
+		if err != nil {
+			return fmt.Errorf("unable to persist ClientSession: %v",
+				err)
+		}
+
+		log.Debugf("New session negotiated with %x@%v, policy: %s",
+			lnAddr.IdentityKey.SerializeCompressed(),
+			lnAddr.Address.String(), clientSession.Policy)
+
+		// We have a newly negotiated session, return it to the
+		// dispatcher so that it can update how many outstanding
+		// negotiation requests we have.
+		select {
+		case n.successfulNegotiations <- clientSession:
+			return nil
+		case <-n.quit:
+			return ErrNegotiatorExiting
+		}
+
+	// TODO(conner): handle error codes properly
+	case wtwire.CreateSessionCodeRejectBlobType:
+		return fmt.Errorf("tower rejected blob type: %v",
+			policy.BlobType)
+
+	case wtwire.CreateSessionCodeRejectRejectMaxUpdates:
+		return fmt.Errorf("tower rejected max updates: %v",
+			policy.MaxUpdates)
+
+	case wtwire.CreateSessionCodeRejectRewardRate:
+		// The tower rejected the session because of the reward rate. If
+		// we didn't request a reward session, we'll treat this as a
+		// permanent tower failure.
+		if !policy.BlobType.Has(blob.FlagReward) {
```

Comment:
Possible that it could, but atm a config change like that would require a restart. In `CheckRemoteInit` we'd ensure that the server supported w/e features we deem necessary. If they say we have a bad reward rate, and the we didn't specify one at all, then it would likely point to a bug in the tower server.